### PR TITLE
Feat/tanlang/add user isolation / 增加用户数据隔离

### PR DESCRIPTION
**Pull request by LinZexiao**
## 关联的Issues (Related Issues)

close https://github.com/filecoin-project/venus/issues/5609
ref https://github.com/filecoin-project/venus/issues/5384

## 自查清单 (Checklist)

在你认为本 PR 满足被审阅的标准之前，需要确保 / Before you mark the PR ready for review, please make sure that:
- [x] 符合Venus项目管理规范中关于PR的[相关标准](https://github.com/ipfs-force-community/dev-guidances/blob/master/%E9%A1%B9%E7%9B%AE%E7%AE%A1%E7%90%86/Venus/PR%E5%91%BD%E5%90%8D%E8%A7%84%E8%8C%83.md) / The PR follows the PR standards set out in the Venus project management guidelines
- [x] 具有清晰明确的[commit message](https://github.com/ipfs-force-community/dev-guidances/blob/master/%E8%B4%A8%E9%87%8F%E7%AE%A1%E7%90%86/%E4%BB%A3%E7%A0%81/git%E4%BD%BF%E7%94%A8/commit-message%E9%A3%8E%E6%A0%BC%E8%A7%84%E8%8C%83.md) / All commits have a clear commit message.
- [x] 包含相关的的[测试用例](https://github.com/ipfs-force-community/dev-guidances/blob/master/%E8%B4%A8%E9%87%8F%E7%AE%A1%E7%90%86/%E4%BB%A3%E7%A0%81/%E4%BB%A3%E7%A0%81%E5%BA%93/%E6%A3%80%E6%9F%A5%E9%A1%B9/%E5%8D%95%E5%85%83%E6%B5%8B%E8%AF%95.md)或者不需要新增测试用例 / This PR has tests for new functionality or change in behaviour or not need to add new tests.
- [x] 包含相关的的指南以及[文档](https://github.com/ipfs-force-community/dev-guidances/tree/master/%E8%B4%A8%E9%87%8F%E7%AE%A1%E7%90%86/%E6%96%87%E6%A1%A3)或者不需要新增文档 / This PR has updated usage guidelines and documentation or not need 
- [ ] 通过必要的检查项 / All checks are green

### 不需要用户隔离的接口
SetSharedParams(ctx context.Context, params *mtypes.SharedSpec) //perm:admin
SetLogLevel(ctx context.Context, subsystem, level string) //perm:admin
SaveNode(ctx context.Context, node *mtypes.Node) //perm:admin
ListNode(ctx context.Context) //perm:admin
HasNode(ctx context.Context, name string) //perm:admin
GetSharedParams(ctx context.Context) //perm:admin
GetNode(ctx context.Context, name string) //perm:admin
DeleteNode(ctx context.Context, name string) //perm:admin
ListMessageByAddress(ctx context.Context, addr address.Address) //perm:admin
ListMessageByFromState(ctx context.Context, from address.Address, state mtypes.MessageState, isAsc bool, pageIndex, pageSize int) //perm:admin
UpdateAllFilledMessage(ctx context.Context) //perm:admin
RepublishMessage(ctx context.Context, id string) //perm:admin
LogList(context.Context) //perm:write
NetPeers //perm:read
NetFindPeer //perm:read
NetAddrsListen //perm:read
NetConnect //perm:admin
UpdateNonce(ctx context.Context, addr address.Address, nonce uint64) //perm:admin
HasMessageByUid(ctx context.Context, id string) //perm:read 

### 已经实现用户隔离的接口

HasAddress(ctx context.Context, addr address.Address) //perm:read
GetMessageByUnsignedCid(ctx context.Context, cid cid.Cid) //perm:read
GetMessageByUid(ctx context.Context, id string) //perm:read
GetMessageBySignedCid(ctx context.Context, cid cid.Cid) //perm:read
GetMessageByFromAndNonce(ctx context.Context, from address.Address, nonce uint64) //perm:read
PushMessageWithId(ctx context.Context, id string, msg *types.Message, meta *mtypes.SendSpec) //perm:write
PushMessage(ctx context.Context, msg *types.Message, meta *mtypes.SendSpec) //perm:write
WalletHas(ctx context.Context, addr address.Address) //perm:read
WaitMessage(ctx context.Context, id string, confidence uint64) //perm:read
UpdateMessageStateByID(ctx context.Context, id string, state mtypes.MessageState) //perm:admin
UpdateFilledMessageByID(ctx context.Context, id string) //perm:admin
SetFeeParams(ctx context.Context, params *mtypes.AddressSpec) //perm:admin
ReplaceMessage(ctx context.Context, params *mtypes.ReplacMessageParams) //perm:admin
RecoverFailedMsg(ctx context.Context, addr address.Address) //perm:admin
MarkBadMessage(ctx context.Context, id string) //perm:admin
ListMessage(ctx context.Context ,  p *types.MsgQueryParams)  //perm:admin   change params
ListFailedMessage(ctx context.Context) //perm:admin
ListBlockedMessage(ctx context.Context, addr address.Address, d time.Duration) //perm:admin
ListAddress(ctx context.Context) //perm:admin
GetAddress(ctx context.Context, addr address.Address) //perm:admin
ForbiddenAddress(ctx context.Context, addr address.Address) //perm:admin
DeleteAddress(ctx context.Context, addr address.Address) //perm:admin
ClearUnFillMessage(ctx context.Context, addr address.Address) //perm:admin
ActiveAddress(ctx context.Context, addr address.Address) //perm:admin
Send(ctx context.Context, params mtypes.QuickSendParams) //perm:sign
SetSelectMsgNum(ctx context.Context, addr address.Address, num uint64) //perm:admin

#### 参数变动
ListMessage(ctx context.Context )  -> ListMessage(ctx context.Context ,  p *types.MsgQueryParams) 

#### 接口权限变动
LogList(context.Context)  //perm:write -> admin 
UpdateMessageStateByID(ctx context.Context, id string, state mtypes.MessageState) //perm:admin -> write
UpdateFilledMessageByID(ctx context.Context, id string) //perm:admin -> write
SetFeeParams(ctx context.Context, params *mtypes.AddressSpec) //perm:admin -> write
ReplaceMessage(ctx context.Context, params *mtypes.ReplacMessageParams) //perm:admin -> write
RecoverFailedMsg(ctx context.Context, addr address.Address) //perm:admin -> write
MarkBadMessage(ctx context.Context, id string) //perm:admin -> write
ListMessage(ctx context.Context ,  p *types.MsgQueryParams)  //perm:admin   -> read
ListFailedMessage(ctx context.Context) //perm:admin -> read
ListBlockedMessage(ctx context.Context, addr address.Address, d time.Duration) //perm:admin -> read
ListAddress(ctx context.Context) //perm:admin -> read
GetAddress(ctx context.Context, addr address.Address) //perm:admin -> read
ForbiddenAddress(ctx context.Context, addr address.Address) //perm:admin -> write
DeleteAddress(ctx context.Context, addr address.Address) //perm:admin -> write
ClearUnFillMessage(ctx context.Context, addr address.Address) //perm:admin ->write
ActiveAddress(ctx context.Context, addr address.Address) //perm:admin ->write
Send(ctx context.Context, params mtypes.QuickSendParams) //perm:admin -> sign
SetSelectMsgNum(ctx context.Context, addr address.Address, num uint64) //perm:admin -> write
NetPeers //perm:read -> admin
NetFindPeer //perm:read -> admin
NetAddrsListen //perm:read -> admin

## 接口测试
与本PR关联的测试主要有两个部分
- 接口变动测试(权限变动,参数变动)
- 用户数据隔离测试

第一部分直接根据变动的细节构造测试用例即可

### 用户数据隔离测试
第二部分需要实现在 msg 中生成至少两个用户的不同数据(包括 消息 ,和地址 )
判断数据是否能被某个用户访问的唯一标准是, 该用户是否具有关联地址的签名权, 也就是该签名地址是否支持测试用户

##### 测试数据

用户1:  admin 权限   
用户2:  sign权限
用户3:   sign权限       

消息1:   由地址1 发送
消息2:   由地址2 发送
消息3:   由地址3 发送

地址1: 支持用户2
地址2: 支持用户2
地址3: 支持用户3

##### 预期结果
用户1 可以访问 所有的数据
用户2  可以访问 地址1,和地址2的关联所有数据,包括发消息的参数和已发送的消息,以及打算发送的消息
用户3  只能访问地址3的关联数据

举例说明:
- GetMessageByUid 时, 
用户1 可以拿到三条消息,
用户2 能够拿到消息1和消息2,
用户3只能拿到消息3

### CODE REVIEW

File: api/messager_impl.go [hunk 1]
```diff
@@ -233,3 +446,49 @@ func (m MessageImp) SetLogLevel(ctx context.Context, subSystem, level string) er
 func (m MessageImp) LogList(ctx context.Context) ([]string, error) {
 	return logging.GetSubsystems(), nil
 }
+
+func (m MessageImp) checkPermissionByAddr(ctx context.Context, addrs ...address.Address) (bool, error) {
```
Comment (simlecode) [inline api/messager_impl.go:450]: 感觉可以直接返回一个error

File: api/messager_impl.go [hunk 2]
```diff
@@ -107,54 +209,153 @@ func (m MessageImp) UpdateAllFilledMessage(ctx context.Context) (int, error) {
...
+	}
+	if m.checkPermissionByName(ctx, msg.WalletName) {
+		return m.MessageSrv.ReplaceMessage(ctx, params)
+	}
+	return cid.Undef, ErrorPermissionDeny
 }
 
 func (m MessageImp) RepublishMessage(ctx context.Context, id string) error {
-	return m.MessageSrv.RepublishMessage(ctx, id)
+	msg, err := m.MessageSrv.GetMessageByUid(ctx, id)
```
Comment (hunjixin) [inline api/messager_impl.go:234]: RepublishMessage接口感觉可以不给用户， replace之后增加一个republish的

File: api/messager_impl.go [hunk 3]
```diff
@@ -35,51 +49,118 @@ func NewMessageImp(implParams ImplParams) *MessageImp {
...
 
+var _ messager.IMessager = (*MessageImp)(nil)
+
 func (m MessageImp) HasMessageByUid(ctx context.Context, id string) (bool, error) {
-	return m.MessageSrv.HasMessageByUid(ctx, id)
+	msg, err := m.MessageSrv.GetMessageByUid(ctx, id)
+	if err != nil {
+		return false, fmt.Errorf("get message by id error: %w", err)
+	}
+	if m.checkPermissionByName(ctx, msg.WalletName) {
```
Comment (hunjixin) [inline api/messager_impl.go:72]: 用FromUser， 按说所有的WaletName都要换成FromUser，walletName/FromUser保持双写， WalletName是历史上的东西，应该是要删除的。

File: api/messager_impl.go (hunk 1 above)
Comment (LinZexiao) [reply api/messager_impl.go:450]: 直接返回 error,通过 有无error 来判断 是否 具有权限确实要比较简练一些,不过我感觉, 将判断结果和 error 分开会不会语义更加清晰一些

File: api/messager_impl.go (hunk 3 above)
Comment (LinZexiao) [reply api/messager_impl.go:72]: 好的

File: api/messager_impl.go [hunk 4]
```diff
@@ -107,10 +204,24 @@ func (m MessageImp) UpdateAllFilledMessage(ctx context.Context) (int, error) {
...
+	}
 	return m.MessageSrv.UpdateFilledMessageByID(ctx, id)
 }
 
 func (m MessageImp) ReplaceMessage(ctx context.Context, params *types.ReplacMessageParams) (cid.Cid, error) {
+	msg, err := m.MessageSrv.GetMessageByUid(ctx, params.ID)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("get message by id error: %w", err)
+	}
+	if check_err := m.checkPermissionBySigner(ctx, msg.From); check_err != nil {
```
Comment (hunjixin) [inline api/messager_impl.go:222]: 同上

File: api/messager_impl.go [hunk 5]
```diff
@@ -119,54 +230,104 @@ func (m MessageImp) RepublishMessage(ctx context.Context, id string) error {
...
 func (m MessageImp) WalletHas(ctx context.Context, addr address.Address) (bool, error) {
+	if err := m.checkPermissionBySigner(ctx, addr); err != nil {
+		return false, err
+	}
 	return m.AddressSrv.WalletHas(ctx, addr)
 }
 
 func (m MessageImp) ListAddress(ctx context.Context) ([]*types.Address, error) {
-	return m.AddressSrv.ListAddress(ctx)
+	msgs, err := m.AddressSrv.ListAddress(ctx)
```
[the remark this answers is not in the document]
Comment (hunjixin) [reply api/messager_impl.go:272]: 保持这样吧，地址反正也比较少

**Comment by hunjixin** [comment]
UpdateFilledMessageByID视线中，如果message的状态已经是Onchain或者是replace， 这不需要再去更新状态了。防止用户，反复点击更新状态。

File: api/messager_impl.go [hunk 6]
```diff
@@ -35,51 +49,117 @@ func NewMessageImp(implParams ImplParams) *MessageImp {
...
+	AuthClient jwtclient.IAuthClient
 }
 
+var _ messager.IMessager = (*MessageImp)(nil)
+
 func (m MessageImp) HasMessageByUid(ctx context.Context, id string) (bool, error) {
-	return m.MessageSrv.HasMessageByUid(ctx, id)
+	msg, err := m.MessageSrv.GetMessageByUid(ctx, id)
+	if err != nil {
+		return false, fmt.Errorf("get message by id error or permission deny")
```
[the remark this answers is not in the document]
Comment (LinZexiao) [reply api/messager_impl.go:70]: 因为这个接口的用途是判断 消息是否存在,如果直接返回错误的话,即使用户没有相关权限也能通过错误信息推断消息是否存在

**Comment by LinZexiao** [comment]
> UpdateFilledMessageByID

已经修改

File: models/mysql/message_test.go [hunk 7]
```diff
@@ -20,6 +20,40 @@ import (
 	"github.com/filecoin-project/venus-messager/testhelper"
 )
 
+func TestListMessageByParams(t *testing.T) {
+	r, mock, sqlDB := setup(t)
+
+	ids := []string{"msg1", "msg2", "msg3"}
```
Comment (hunjixin) [inline models/mysql/message_test.go:26]: 删除返回值部分的check， 没有什么意义

File: models/mysql/message_test.go [hunk 8]
```diff
@@ -341,29 +373,68 @@ func testListMessageByAddress(t *testing.T, r repo.Repo, mock sqlmock.Sqlmock) {
...
+			WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(ids[0]).AddRow(ids[1]))
+
+		res, err := r.MessageRepo().ListFailedMessage(&repo.MsgQueryParams{State: []types.MessageState{types.OnChainMsg}})
+		assert.NoError(t, err)
+		checkMsgWithIDs(t, res, ids)
+	})
+
+	t.Run("indicate from", func(t *testing.T) {
+		addr := testutil.AddressProvider()(t)
+		mock.ExpectQuery(regexp.QuoteMeta("SELECT * FROM `messages` WHERE `from_addr` = ? AND `state` = ? AND (error_msg is not null) ORDER BY created_at")).
```
Comment (hunjixin) [inline models/mysql/message_test.go:400]: 可能要补充一个in的例子

File: api/mesager_impl_test.go [hunk 9]
```diff
@@ -0,0 +1,800 @@
...
+	"github.com/filecoin-project/venus-auth/jwtclient"
+	"github.com/filecoin-project/venus-messager/mocks"
+	"github.com/filecoin-project/venus-messager/testhelper"
+	types "github.com/filecoin-project/venus/venus-shared/types/messager"
+	"github.com/golang/mock/gomock"
+	"github.com/ipfs/go-cid"
+	"github.com/stretchr/testify/assert"
+)
+
+//go:generate mockgen -destination=../mocks/mock_auth_client.go -package=mocks github.com/filecoin-project/venus-auth/jwtclient IAuthClient
```
Comment (hunjixin) [inline api/mesager_impl_test.go:21]: 这个是没用的吧

File: models/mysql/message_test.go (hunk 8 above)
Comment (LinZexiao) [reply models/mysql/message_test.go:400]: 已经补充了

File: testhelper/mock_authclient.go [hunk 10]
```diff
@@ -75,6 +75,40 @@ func (m *AuthClient) AddMockUserAndSigner(account string, addrs []address.Addres
 	}
 }
 
+func (m *AuthClient) HasUser(req *auth.HasUserRequest) (bool, error) {
```
[the remark this answers is not in the document]
Comment (LinZexiao) [reply testhelper/mock_authclient.go:78]: 可以在另一个PR里面修改

File: models/mysql/message_test.go (hunk 7 above)
Comment (LinZexiao) [reply models/mysql/message_test.go:26]: 已删除

**Review by simlecode — APPROVED** [review]
LGTM

File: utils/util.go [hunk 11]
```diff
@@ -124,3 +129,24 @@ func MsgsGroupByAddress(msgs []*types.SignedMessage) map[address.Address][]*type
...
+	}
+	tk, err := jwt.Sign(pl, jwt.NewHS256(secret))
+	if err != nil {
+		return core.EmptyString, xerrors.Errorf("gen token failed :%s", err)
+	}
+	return string(tk), nil
+}
+
+// RandSecret If the daemon does not have a secret key configured, it is automatically generated
+func RandSecret() ([]byte, error) {
```
Comment (hunjixin) [inline utils/util.go:146]: venus-auth里面有这个

File: models/repo/node_repo.go [hunk 12]
```diff
@@ -6,6 +6,8 @@ type INodeProvider interface {
 	ListNode() ([]*types.Node, error)
 }
 
+//go:generate  mockgen -destination=../../mocks/mock_node_repo.go -package=mocks github.com/filecoin-project/venus-messager/models/repo NodeRepo
```
Comment (hunjixin) [inline models/repo/node_repo.go:9]: 1

Comment (hunjixin) [reply models/repo/node_repo.go:9]: 1

File: utils/util.go (hunk 11 above)
Comment (LinZexiao) [reply utils/util.go]: 已经复用 authclient 中的 RandSecret

**Review by hunjixin — APPROVED** [review]
LGR

**Comment by hunjixin** [comment]
@LinZexiao  麻烦增加一个测试说明

**Pull request merged**